File: C/0007.c
#include <stdio.h>
#include <stdbool.h>

// Function to check if a number is prime
bool is_prime(int num) {
    if (num <= 1) return false;
    if (num == 2) return true;
    if (num % 2 == 0) return false;

    for (int i = 3; i * i <= num; i += 2) {
        if (num % i == 0) return false;
    }
    return true;
}

int main() {
    int count = 0;
    int number = 1;
    int target = 10001;

    // Find the 10001st prime number
    while (count < target) {
        number++;
        if (is_prime(number)) {
            count++;
        }
    }

    printf("The %dth prime number is %d\n", target, number);

    return 0;
}

File: TypeScript/0007.ts
function isPrime(num: number, primes: number[]): boolean {
    const sqrt = Math.sqrt(num);
    for (const prime of primes) {
        if (prime > sqrt) break; // No need to check beyond sqrt(num)
        if (num % prime === 0) return false;
    }
    return true;
}

function nthPrime(n: number): number {
    if (n === 1) return 2; // The first prime is 2

    const primes: number[] = [2]; // List to store primes
    let candidate = 3; // Start checking from 3

    while (primes.length < n) {
        if (isPrime(candidate, primes)) {
            primes.push(candidate); // Add prime to the list
        }
        candidate += 2; // Increment by 2 to check only odd numbers
    }

    return primes[n - 1]; // Return the nth prime
}

// Usage
const n = 10001;
console.log(`The ${n}st prime number is:`, nthPrime(n));